File: backend/app/schemas/reviews.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator


class ReviewMediaItem(BaseModel):
    type: Literal['image', 'video']
    url: str
    thumbnail_url: str | None = None
    alt: str | None = None


class ReviewBase(BaseModel):
    rating: int
    title: str | None = None
    body: str
    media: list[ReviewMediaItem] = Field(default_factory=list)

    @field_validator('rating')
    @classmethod
    def validate_rating(cls, v: int) -> int:
        if not (1 <= v <= 5):
            raise ValueError('Rating must be between 1 and 5')
        return v


class ReviewCreate(ReviewBase):
    product_id: str | None = None


class ReviewUpdate(BaseModel):
    title: str | None = None
    body: str | None = None
    media: list[ReviewMediaItem] | None = None


class ReviewModeration(BaseModel):
    status: Literal['pending', 'approved', 'rejected']
    moderation_comment: str | None = None


class ReviewResponse(BaseModel):
    response: str = Field(..., min_length=1, description='Response text to the review')


class Review(ReviewBase):
    id: str
    organization_id: str
    product_id: str | None = None
    author_user_id: str
    status: Literal['pending', 'approved', 'rejected']
    moderated_by: str | None = None
    moderated_at: datetime | None = None
    moderation_comment: str | None = None
    response: str | None = None
    response_by: str | None = None
    response_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PublicReview(BaseModel):
    id: str
    product_id: str | None = None
    author_user_id: str
    rating: int
    title: str | None = None
    body: str
    media: list[ReviewMediaItem] = Field(default_factory=list)
    response: str | None = None
    response_at: datetime | None = None
    created_at: datetime


class ReviewsResponse(BaseModel):
    items: list[Review]
    total: int


class PublicReviewsResponse(BaseModel):
    items: list[PublicReview]
    total: int
    average_rating: float | None = None


class ReviewStats(BaseModel):
    total_reviews: int
    average_rating: float | None = None
    rating_distribution: dict[int, int] = Field(default_factory=dict)  # {1: count, 2: count, ...}

File: frontend/src/types/reviews.ts
export interface ReviewMediaItem {
  type: 'image' | 'video'
  url: string
  thumbnail_url?: string | null
  alt?: string | null
}

export interface Review {
  id: string
  organization_id: string
  product_id?: string | null
  author_user_id: string
  rating: number
  title?: string | null
  body: string
  media: ReviewMediaItem[]
  status: 'pending' | 'approved' | 'rejected'
  moderated_by?: string | null
  moderated_at?: string | null
  moderation_comment?: string | null
  response?: string | null
  response_by?: string | null
  response_at?: string | null
  created_at: string
  updated_at: string
}

export interface ReviewCreate {
  product_id?: string | null
  rating: number
  title?: string | null
  body: string
  media?: ReviewMediaItem[]
}

export interface ReviewUpdate {
  title?: string | null
  body?: string | null
  media?: ReviewMediaItem[]
}

export interface ReviewModeration {
  status: 'pending' | 'approved' | 'rejected'
  moderation_comment?: string | null
}

export interface ReviewResponse {
  response: string
}

export interface PublicReview {
  id: string
  product_id?: string | null
  author_user_id: string
  rating: number
  title?: string | null
  body: string
  media: ReviewMediaItem[]
  response?: string | null
  response_at?: string | null
  created_at: string
}

export interface ReviewsResponse {
  items: Review[]
  total: number
}

export interface PublicReviewsResponse {
  items: PublicReview[]
  total: number
  average_rating?: number | null
}

export interface ReviewStats {
  total_reviews: number
  average_rating?: number | null
  rating_distribution: {
    1: number
    2: number
    3: number
    4: number
    5: number
  }
}

// AI Response Generation Types
export type ResponseTone = 'professional' | 'friendly' | 'apologetic'
export type ReviewSentiment = 'positive' | 'neutral' | 'negative'

export interface AIResponseSuggestion {
  tone: ResponseTone
  text: string
  confidence: number
}

export interface AIResponseResult {
  sentiment: ReviewSentiment
  topics: string[]
  suggestions: AIResponseSuggestion[]
}

